### DISCUSSION TITLE
Average Directional Movement Index (ADX)

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#270] [Indicators]
Community discussion about [Average Directional [Movement] Index (ADX)](https://dotnet.stockindicators.dev/indicators/Adx/#content)

**Comment by bobtabor** [comment]
Is there an equivalent indicator focused on *volume*?  Or any indicator where I can simple see whether volume is increasing or decreasing over, say, the last 5 or 10 minutes?

**Reply by DaveSkender** [reply]
If you're looking for something like ADX, [Price Volume Oscillator](https://daveskender.github.io/Stock.Indicators/indicators/Pvo/#content) is an interesting one to explore.

Generally, there are a few indicators in the library that use volume:
https://dotnet.stockindicators.dev/indicators/#volume-based

The simplest is merely a [moving average of volume](https://dotnet.stockindicators.dev/indicators/Sma/#content), where you can compare current price to the rolling average (over or under).

**Reply by DaveSkender** [reply]
[Slope](https://daveskender.github.io/Stock.Indicators/indicators/Slope/#content) of volume might interesting too.

**Comment by bobtabor** [comment]
re: slope

Yes, that is what I am doing now.  I was hoping to translate slope into something more meaningful because I can't make sense of the values other than "positive" or "negative".

I was doing this to turn the slope into degrees, hoping it could help me visualize the extent of the upward or downward momentum.  This is what I was doing:

```
var result = LinearRegression.Calculate(points);
var degrees = Math.Atan(result.Slope) * (180.0 / Math.PI);
```

But in one case, I was getting a -89 ... it was then I realized that the scale is important when considering slope.  :/

**Reply by DaveSkender** [reply]
Yah, might be easier to do the slope of a small moving average to filter the noise.

**Reply by tgvolk** [reply]
this may help converting slope to degrees.  Figure an average size monitor (results may vary), but I can get close to the correct degrees with this:
```
var slopeAdjustment = (zedGraphControl1.GraphPane.YAxis.Scale.Max - zedGraphControl1.GraphPane.YAxis.Scale.Min) / (numPer * .25);
var degrees = (Math.Atan(currentSlope / slopeAdjustment ) * (180 / Math.PI);
```

**Reply by DaveSkender** [reply]
It's a bit of a tangled mess when you get into the circular logic needed to get the geometry right.  I tried once, but gave up after a while as it wasn't worth the effort for my use case.

It all gets back to the aspect ratio of your chart plot, how many bars of data you're showing, _and_ the axis scales.  Throw in responsive user interface design and it pushes you towards an entangled software architecture where the data tier relies on the user interface tier.

With that said, I think it's possible, but you'd probably need to either enforce fixed values for: chart aspect ratio, chart axis size, and number of bars plotted; or, let the user interface dynamically calculate % values as the screen size changes.

**Comment by hal9000cc** [comment]
I discovered a very strange thing. ADX on binance does not use TR/ATR in the calculation. Your indicator also does not use TR/ATR in calculating the ADX value and the result is the same as binance. But in all the sources that I found, the algorithm is different. It uses TR or ATR.

**Reply by DaveSkender** [reply]
We use it, somewhat.  These are represented by `tr` and `trs` variables, respectively [in the source code](https://github.com/DaveSkender/Stock.Indicators/blob/6d403309ddb60ea74f3619d5a07327dca48282c5/src/a-d/Adx/Adx.Series.cs) if you want to look under the hood.

The original author, Welles Wilder, uses a smoothed sum of TR, so it’s not technically ATR.  Using ATR probably still works, since ADX is really a formula of ratios, but it adds unnecessary computational workloads.

**Reply by DaveSkender** [reply]
I learned a while ago that there’s a lot of misleading explanations out there, and a lot of bad code.  You’d be surprised at how many unreliable sources and charting sites there are, who just copy and paste bad info from each other.  I always source from original author explanations and mathematical formulas whenever possible, and never use other people’s code as sources of truth.

**Reply by hal9000cc** [reply]
I looked at the source code. You use tr values, but they don't affect the ADX result. They only affect the Mdi and Pdi values. Try to remove tr from the ADX calculation, and its result will not change.

Here is a simple math: abs(100 * pdm / trs - 100 * mdm/ trs) / (100 * pdm/ trs + 100 * mdm/trs) = abs(pdm - mdm) / (pdm + mdm)

**Reply by hal9000cc** [reply]
I'm making my ADX indicator on python. This code outputs values that are exactly the same as yours, and it does not use TR/ATR:

```py 
    high = ohlcv.high
    low = ohlcv.low

    p_dm = np.hstack([np.nan, np.diff(high)])
    m_dm = np.hstack([np.nan, -np.diff(low)])

    bx_zero_plus_dm = (p_dm <= m_dm) | (p_dm < 0)
    bx_zero_minus_dm = (m_dm <= p_dm) | (m_dm < 0)
    p_dm[bx_zero_plus_dm] = 0
    m_dm[bx_zero_minus_dm] = 0

    p_di = ma_calculate(p_dm, period, ma_type_enum)
    m_di = ma_calculate(m_dm, period, ma_type_enum)

    dxi = 100 * abs(p_di - m_di) / (p_di + m_di)
    dxi[p_di + m_di == 0] = 0

    adx = ma_calculate(dxi, smooth, ma_type_enum)
```

**Reply by hal9000cc** [reply]
I think there is a mistake somewhere somewhere....

**Reply by DaveSkender** [reply]
Oh, I see what you’re saying.  You’re trying to see if it’s possible that Wilder’s method can be reduced to a simpler math for ADX only, without errors.  I’d have to do a full mathematical proof to know for sure.

**Reply by DaveSkender** [reply]
Based on a quick look, it looks like your simplification follows the basic distributive properties of math correctly, using `trs` in this snippet does not change `dx` or it’s average.

``` csharp
// directional increments
pdi = 100 * pdm / trs;
mdi = 100 * mdm / trs;

// calculate ADX
dx = … 100 * Math.Abs(pdi - mdi) / (pdi + mdi)
[..]
```

As long as you don’t need correctly scaled DI+ and DI- values, this is probably an okay shortcut to ADX.  We’re producing all of these, so it wouldn’t be appropriate for us to remove TR

**Reply by hal9000cc** [reply]
Yes, what proofs, ordinary school algebra.
(A / X - B / X) / (A / X + B / X) = ((A - B) / X) / ((A + B) / X) = (A - B) / (A + B)

I think we are not counting correctly, so TR is being reduced. I don't think Wilder couldn't cut a simple formula :)

We calculate like this:
    p_di = ma_calculate(p_dm, period, ma_type_enum) / tr
    m_di = ma_calculate(m_dm, period, ma_type_enum) / tr

I will assume that it is necessary so:
    p_di = ma_calculate(p_dm / tr, period, ma_type_enum)
    m_di = ma_calculate(m_dm / tr, period, ma_type_enum)

In this case, nothing is reduced.
I don't have access to the original source to verify this. I'll keep the second option for myself.

**Reply by hal9000cc** [reply]
But I have doubts. If you have a link to the original source, I will be grateful for it.

**Reply by DaveSkender** [reply]
He uses TR to normalize the positive and negative incremental directional movements, DI+ and DI-.  These are separate but still useful data elements.  With your simplification the red and green lines in this family of indicators would be denormalize (not aligned with author’s intent).

![6DF52B96-259A-4127-B0BF-66D2B5CD15DD](https://user-images.githubusercontent.com/8432125/203835631-4156f5aa-e987-49bc-83c7-c15ea7433447.jpeg)

**Reply by DaveSkender** [reply]
Wilder describes this in his book “New Concepts in Technical Trading Systems”.  It’s an actual book, so I can’t send you a link.

**Reply by DaveSkender** [reply]
If you want to save yourself some time, consider just using our [Python wrapper](https://python.stockindicators.dev/indicators/Adx/#content).  We've done all this work for you.

**Comment by hal9000cc** [comment]
Good day!
I can't figure out how the moving average is calculated. Adx.Series.cs:
```
            if (i == lookbackPeriods)
            {
                trs = sumTr;
                pdm = sumPdm;
                mdm = sumMdm;
            }
            else
            {
                trs = prevTrs - (prevTrs / lookbackPeriods) + tr;
                pdm = prevPdm - (prevPdm / lookbackPeriods) + pdm1;
                mdm = prevMdm - (prevMdm / lookbackPeriods) + mdm1;
            }
```

Are you sure it shouldn't be like this?
```
            if (i == lookbackPeriods)
            {
                trs = sumTr;
                pdm = sumPdm;
                mdm = sumMdm;
            }
            else
            {
                trs = prevTrs - (prevTrs / lookbackPeriods) + (tr / lookbackPeriods);
                pdm = prevPdm - (prevPdm / lookbackPeriods) + (pdm1 / lookbackPeriods);
                mdm = prevMdm - (prevMdm / lookbackPeriods) + (mdm1 / lookbackPeriods);
            }
```

**Reply by DaveSkender** [reply]
I can see how this may not be very clear since it's a multi-step process to get to the final ADX values.  What you're showing here is both the initial seed and raw, rolling, cumulative, intermediate values of:

- true range
- positive directional movement
- negative directional movement

To get to the `index` and averaging parts, you have to further process this raw information with a smoothed moving average method.  Including the snippets I left you above, you'd then proceed to:

```csharp
adx = ((prevAdx * (lookbackPeriods - 1)) + dx) / lookbackPeriods;
```

**Reply by hal9000cc** [reply]
Yes, I see both stages of smoothing. But in the first stage, in order to get ATR, and the smoothed moment of positive and negative directional movement, it is necessary to add '/ lookbackPeriods' in the formula to get an exponential moving average with a = 1 / N. 

What is happening there now, I don't understand at all. I think there is still a mistake, check it out.

**Reply by DaveSkender** [reply]
This is the method for averaging, using the SMMA equation.  The `prevPdm`, for example, has already been averaged in the prior iteration, so you'd not need to divide by periods again to "average".  Rather, its use is to simply take the ratio of it and add it to the newer `pdm1` increment.

**Reply by DaveSkender** [reply]
If you want to save yourself some time, consider just using our [Python wrapper](https://python.stockindicators.dev/indicators/Adx/#content).  We've done all this work for you.

**Reply by hal9000cc** [reply]
That's right, SMMA is an EMA with coefficient a=1/N. The general formula is
S(t) = a * x(t) + (1-a) * S(t-1).
If we substitute a=1/N, then this formula can be written as you use:
S(t) = 1/N * x(t) + (1 - 1/N) * S(t-1) = x(t) / N + S(t-1) - S(t-1) / N = S(t-1) - S(t-1) / N + x(t) / N
You are missing one action.

You can compare the ATR values with your indicator and see that they differ. If you add '/N', the values will match.

**Reply by hal9000cc** [reply]
By the way, due to the reduction of values, this error does not affect the ATR, it is correct. But DI+ and DI- turn out to be erroneous.

**Reply by DaveSkender** [reply]
I'm going to try to understand what you're saying for the sake of ensuring that we're doing it correctly.  I'm also trying to be careful to be true to the original author's methods.  What's confusing me, is that it looks like you're trying to use someone else's method.

**Reply by hal9000cc** [reply]
You have a wonderful library of indicators. I'm writing my library and using yours for validation. I find errors in my library. But now I seem to have found in yours :)
I found the book by J. Welles Wilder. Everything you have corresponds to the author's methodology. It's just an error when smoothing in the formula :)
Just compare the ATR values and you will see this.

**Reply by DaveSkender** [reply]
I'm not yet convinced you've found an error!  He literally doesn't use ATR in the normal sense, so that's where I think there's a disconnect.  He's using a sum of TR, which is different.  Given the same book defines ATR, he would have explicitly called out ATR in his description of ADX if that was his intent.  He does not.

**Reply by DaveSkender** [reply]
What's confusing in Wilder's explanation is this use of TR<sub>14</sub>.  It's initialized as the sum of daily TR, then is maintained with an SMMA smoothing technique by adding increments of daily TR values.  For his 14 period example, he writes:

> TR<sub>14</sub>(today) = Previous TR<sub>14</sub> - (Previous TR<sub>14</sub>/14) + TR<sub>1</sub>

which is exactly what I have:

> trs = prevTrs - (prevTrs / lookbackPeriods) + tr;

If you read this carefully, it is not an average of TR, but rather a sum of TR over the lookback periods.

**Reply by hal9000cc** [reply]
Now I thought - what will be the difference between the curves, if there is an additional division or not. But in fact it does not affect their shape. Only the amplitude and location will change.

**Reply by hal9000cc** [reply]
Yes, you are right, you have fully implemented the author's method. And there it is considered peculiar. Although, in fact, it is the same ATR, only, it seems, just multiplied by the period.

**Reply by DaveSkender** [reply]
Dividing TR<sub>1</sub> by a size factor here is eventually going to move TR<sub>14</sub> to a smoothed average of TR.  If this were intended to be ATR, that would be correct.  This is supposed to be "total True Range", not "average True Range", so it'd be something else.

And simply multiplying `ATR` by `lookbackPeriods` to reconstitute it to total TR is a bit backward, but I'm sure is a hack some have used.  I can't do the math in my head, but it may produce a very similar but still deviant result.

**Reply by hal9000cc** [reply]
There's another interesting point. All three values count that way. And they are all affected by this effect, and they are divided into each other.
As a result, after a while they begin to completely coincide with my values, which I calculated using the usual MMA smoothing. Here are the last 10 yours and mine:
'''
(array([ 4.31273032,  2.69024009,  9.05587651,  2.55794453,  1.09489647,
        0.63638118,  7.9565679 ,  8.11709597,  4.41693475, 12.82305828]),
array([ 4.31273032,  2.69024009,  9.05587651,  2.55794453,  1.09489647,
        0.63638118,  7.9565679 ,  8.11709597,  4.41693475, 12.82305828]))
'''

**Reply by DaveSkender** [reply]
I'm sure it's possible to reorganize the maths on this one to arrive at the same result.
If you're building your own version of this, instead of using this library, It's up to you to determine if alternative approaches are appropriate for your use cases.

My preference is to be as unopinionated as possible with this library, which is why I always go back to the original authors and never use other people's code or less reputable sources.

**Comment by hal9000cc** [comment]
First 20 values (up - my):

```
(array([        nan,         nan,  0.        ,  0.        ,  0.78914315,
       84.37151483, 65.44538665, 44.95463786, 33.39600899, 24.98930754,
       16.28827223,  7.58056838,  4.00331537, 25.41011753, 15.4902397 ,
       13.50497211, 19.72029074, 12.84137277,  5.74302962, 14.39915213]),
array([        nan,         nan,  0.        ,  0.        ,  0.86371074,
       85.083874  , 65.73148531, 45.08944494, 33.47034823, 25.02843565,
       16.30488697,  7.58416513,  4.00431825, 25.41363659, 15.49154739,
       13.50554353, 19.72076957, 12.84153667,  5.7430624 , 14.39918878]))
```

**Reply by hal9000cc** [reply]
I apologize for the trouble, but I discovered something new :)

**Reply by hal9000cc** [reply]
It is also curious why they differ in the beginning...

**Reply by hal9000cc** [reply]
But they should not differ if the first value matches. So it doesn't match.

**Reply by DaveSkender** [reply]
Consider using our [Python wrapper](https://python.stockindicators.dev/indicators/Adx/#content) instead of reinventing the wheel, to avoid all of your extra work here.  We've done all this work and troubleshooting for you and have 1,000s of users who've live-tested these things over the last 2 years (in addition to our own rigorous testing).  It'd be more efficient to simply contribute to our library if you see or think something is wrong and want to fix it.

**Reply by hal9000cc** [reply]
There is an assumption. It looks like you have at the beginning the sum is not lookbackPeriods, but lookbackPeriods-1. I will still find a mistake with you! :)))

**Reply by DaveSkender** [reply]
Hehe, sounds like you're on a mission :smile:

**Reply by DaveSkender** [reply]
There's 2 things to consider with the starting values.  1) we're using a zero-based index, so `lookbackPeriods - 1` is actually the same bar position as `lookbackPeriods` and 2) The first TR can only be calculated on the second bar; therefore, the first SMA of TR, the seeded value of TR<sub>14</sub> can only happen on the 15th bar, for example.  Our [manual calculations spreadsheet](https://github.com/DaveSkender/Stock.Indicators/blob/024583554eede9460bc22892c20873a21488fab5/tests/indicators/a-d/Adx/Adx.Calc.xlsx) might illuminate some things here for you.

**Reply by hal9000cc** [reply]
It seems to be ok everywhere. Perhaps I am not initializing correctly, I will search.

**Reply by hal9000cc** [reply]
I found out the reason. My ATR is counted separately, and initialized one step earlier. You are calculated in exact accordance with the original method.

**Reply by DaveSkender** [reply]
On a related note, check that you’re using v2.4.1 or newer.  I’d recently updated all True Range related indicators so they don’t try to calculate TR on the first bar with limited information.  Williams is inconsistent about whether or not this is a good idea.  The change did impact ADX, by shifting the first calculable bar by one position.

In other words, since you’re eagerly trying to find problems, I’ll give you partial credit for finding:

> There is an assumption. It looks like you have at the beginning the sum is not lookbackPeriods

However, most credit goes to a [different conversation](https://github.com/DaveSkender/Stock.Indicators/issues/947) with @mihakralj.
